modules/core/src/main/scala/fs2/kafka/KafkaProducer.scala

```diff
@@ -106,28 +106,25 @@ object KafkaProducer {
     valueSerializer: Serializer[F, V],
     producer: KafkaByteProducer
   )(
-    implicit F: ConcurrentEffect[F]
+    implicit F: Concurrent[F]
   ): ProducerRecord[K, V] => F[F[(ProducerRecord[K, V], RecordMetadata)]] =
     record =>
       asJavaRecord(keySerializer, valueSerializer, record).flatMap { javaRecord =>
-        Deferred[F, Either[Throwable, (ProducerRecord[K, V], RecordMetadata)]].flatMap { deferred =>
-          F.delay {
-              producer.send(
-                javaRecord,
-                callback { (metadata, exception) =>
-                  val complete =
-                    deferred.complete {
-                      if (exception == null)
-                        Right((record, metadata))
-                      else Left(exception)
-                    }
-
-                  F.runAsync(complete)(_ => IO.unit).unsafeRunSync()
+        F.async { (cb: Either[Throwable, (ProducerRecord[K, V], RecordMetadata)] => Unit) =>
+            producer.send(
+              javaRecord,
+              callback { (metadata, exception) =>
+                cb {
+                  if (exception == null)
+                    Right((record, metadata))
+                  else Left(exception)
                 }
-              )
-            }
-            .as(deferred.get.rethrow)
-        }
+              }
+            )
+            ()
+          }
```

Comment:
this is CE2, so you need a `guarantee(context.shift)` here

Reply:
although maybe it doesn't matter given that you `start`. Might want to test it if you can be bothered

Reply:
I've put it in the wrong place now, right?